Add more FV functor heat conduction capability

GiudGiud opened the pull request:
refs https://github.com/idaholab/moose/issues/30539

GiudGiud committed:
Add functor support to the heat conduction FV Physics
Add a functor kernel for time derivative term in the heat equation with FV
closes #30539

GiudGiud committed:
Add a utility for checking that two parameters are not compatible

GiudGiud committed:
Add a test for functor support in FV heat conduction physics
refs #30539

GiudGiud committed:
Add an option to use a different names for the functors defined by a general functor fluid props

lindsayad commented on modules/heat_transfer/doc/content/source/fvkernels/FVFunctorHeatConductionTimeDerivative.md:
```diff
@@ -0,0 +1,19 @@
+The `FVFunctorHeatConductionTimeDerivative` kernel implements the residual
+
+\begin{equation}
+\int_{\Omega_C} \rho c_p \frac{\partial u}{\partial t} dV
```
```suggestion
\int_{\Omega_C} \rho c_p \frac{\partial T}{\partial t} dV
```

lindsayad commented on modules/heat_transfer/doc/content/source/fvkernels/FVFunctorHeatConductionTimeDerivative.md:
```diff
@@ -0,0 +1,19 @@
+\end{equation}
+
+where $\rho$ is the material density, $c_p$ is the specific heat capacity at
+constant pressure and $u$ the temperature variable. It uses [functors](Functors/index.md)
```
```suggestion
constant pressure and $T$ the temperature variable. It uses [functors](Functors/index.md)
```

lindsayad commented on modules/heat_transfer/doc/content/source/fvkernels/FVFunctorHeatConductionTimeDerivative.md:
```diff
@@ -0,0 +1,19 @@
+
+where $\rho$ is the material density, $c_p$ is the specific heat capacity at
+constant pressure and $u$ the temperature variable. It uses [functors](Functors/index.md)
+for these properties. The [FVHeatConductionTimeDerivative.md] can be used with material properties instead.
```
```suggestion
for these properties. [FVHeatConductionTimeDerivative.md] can be used with [traditional material properties](/Materials/index.md) instead.
```

lindsayad commented on modules/heat_transfer/doc/content/source/fvkernels/FVHeatConductionTimeDerivative.md:
```diff
@@ -11,6 +11,10 @@ The `FVHeatConductionTimeDerivative` kernel implements the residual
 
+!alert note
+The [FVFunctorHeatConductionTimeDerivative.md] can be used to use functors for these properties instead
+of material properties.
```
```suggestion
of traditional material properties.
```

lindsayad commented on modules/heat_transfer/include/fvkernels/FVFunctorHeatConductionTimeDerivative.h:
```diff
@@ -0,0 +1,29 @@
+
+#include "FVTimeKernel.h"
+
+class FVFunctorHeatConductionTimeDerivative : public FVTimeKernel
```
doxygen plz

lindsayad replied:
```suggestion
class FVFunctorHeatConductionTimeDerivative : public FVFunctorTimeKernel
```

lindsayad commented on modules/heat_transfer/include/fvkernels/FVFunctorHeatConductionTimeDerivative.h:
```diff
@@ -0,0 +1,29 @@
+protected:
+  virtual ADReal computeQpResidual() override;
+
+  /// Specific heat material property
```
```suggestion
  /// Specific heat capacity at constant pressure
```

lindsayad commented on modules/heat_transfer/include/fvkernels/FVFunctorHeatConductionTimeDerivative.h:
```diff
@@ -0,0 +1,29 @@
+  /// Specific heat material property
+  const Moose::Functor<ADReal> & _specific_heat;
+
+  /// Density material property
```
```suggestion
  /// Density
```

lindsayad commented on modules/heat_transfer/src/fvkernels/FVFunctorHeatConductionTimeDerivative.C:
```diff
@@ -0,0 +1,45 @@
+      "AD Time derivative term $\\rho c_p \\frac{\\partial T}{\\partial t}$ of "
+      "the heat equation for quasi-constant specific heat $c_p$ and the density $\\rho$.");
+  params.addParam<MooseFunctorName>(
+      "specific_heat", "specific_heat", "Property name of the specific heat material property");
```
```suggestion
      "specific_heat", "specific_heat", "Property name of the specific heat at constant pressure");
```

lindsayad commented on modules/heat_transfer/src/fvkernels/FVFunctorHeatConductionTimeDerivative.C:
```diff
@@ -0,0 +1,45 @@
+  params.addParam<MooseFunctorName>(
+      "specific_heat", "specific_heat", "Property name of the specific heat material property");
+  params.addParam<MooseFunctorName>(
+      "density", "density", "Property name of the density material property");
```
```suggestion
      "density", "density", "Property name of the density");
```

lindsayad commented on modules/heat_transfer/src/fvkernels/FVFunctorHeatConductionTimeDerivative.C:
```diff
@@ -0,0 +1,45 @@
+ADReal
+FVFunctorHeatConductionTimeDerivative::computeQpResidual()
+{
+  // FV kernels dont use the quadrature
+  mooseAssert(_q_point.size() == 1, "Only 1 QP per element in FV");
```
This is an overzealous assertion at this point. This would correctly evaluate the finite volume term even if there were multiple quadrature points in the quadrature rule unlike the qp-based term implementation

lindsayad replied:
Well this will be true after switching to use `FunctorTimeKernel` as the base class

GiudGiud replied:
followed the change

lindsayad replied:
You're never supposed to use a qrule in an FV kernel

lindsayad replied:
Consequently, there is no loop over qrule qpoints in the base class like you say
```
void
FVElementalKernel::computeResidual()
{
  prepareVectorTag(_assembly, _var.number());
  _local_re(0) += MetaPhysicL::raw_value(computeQpResidual()) * _assembly.elemVolume();
  accumulateTaggedLocalResidual();
}
```

GiudGiud replied:
right I agree with that

lindsayad commented on modules/heat_transfer/src/fvkernels/FVFunctorHeatConductionTimeDerivative.C:
```diff
@@ -0,0 +1,45 @@
+{
+  // FV kernels dont use the quadrature
+  mooseAssert(_q_point.size() == 1, "Only 1 QP per element in FV");
+  const Moose::ElemArg elem_arg = {_current_elem, true};
```
```suggestion
  const Moose::ElemArg elem_arg = {_current_elem, /*correct_skewness=*/true};
```

lindsayad commented on modules/heat_transfer/src/fvkernels/FVFunctorHeatConductionTimeDerivative.C:
```diff
@@ -0,0 +1,45 @@
+  const Moose::ElemArg elem_arg = {_current_elem, true};
+  const auto & time_arg = determineState();
+  return _specific_heat(elem_arg, time_arg) * _density(elem_arg, time_arg) *
+         FVTimeKernel::computeQpResidual();
```
```suggestion
         FVFunctorTimeKernel::computeQpResidual();
```

lindsayad commented on modules/heat_transfer/src/fvkernels/FVFunctorHeatConductionTimeDerivative.C:
```diff
@@ -0,0 +1,45 @@
+InputParameters
+FVFunctorHeatConductionTimeDerivative::validParams()
+{
+  InputParameters params = FVTimeKernel::validParams();
```
```suggestion
  InputParameters params = FVFunctorTimeKernel::validParams();
```

lindsayad commented on modules/heat_transfer/src/fvkernels/FVFunctorHeatConductionTimeDerivative.C:
```diff
@@ -0,0 +1,45 @@
+
+FVFunctorHeatConductionTimeDerivative::FVFunctorHeatConductionTimeDerivative(
+    const InputParameters & parameters)
+  : FVTimeKernel(parameters),
```
```suggestion
  : FVFunctorTimeKernel(parameters),
```

lindsayad commented on modules/heat_transfer/include/fvkernels/FVFunctorHeatConductionTimeDerivative.h:
```diff
@@ -0,0 +1,29 @@
+
+#pragma once
+
+#include "FVTimeKernel.h"
```
```suggestion
#include "FVFunctorTimeKernel.h"
```

lindsayad commented on modules/heat_transfer/src/physics/HeatConductionFV.C:
```diff
@@ -26,11 +26,15 @@ HeatConductionFV::validParams()
+                                            "Thermal conductivity functor material property");
+  params.addParam<MaterialPropertyName>("specific_heat", "Specific heat material property");
+  params.addParam<MooseFunctorName>("specific_heat_functor",
+                                    "Specific heat functor material property");
```
```suggestion
                                    "Specific heat functor");
```

lindsayad commented on modules/heat_transfer/src/physics/HeatConductionFV.C:
```diff
@@ -26,11 +26,15 @@ HeatConductionFV::validParams()
+  params.addParam<MooseFunctorName>("specific_heat_functor",
+                                    "Specific heat functor material property");
+  params.addParam<MaterialPropertyName>("density", "Density material property");
+  params.addParam<MooseFunctorName>("density_functor", "Density functor material property");
```
```suggestion
  params.addParam<MooseFunctorName>("density_functor", "Density functor");
```

lindsayad commented on modules/heat_transfer/src/physics/HeatConductionFV.C:
```diff
@@ -92,11 +104,22 @@ HeatConductionFV::addFVKernels()
   if (isTransient())
   {
-    const std::string kernel_type = "FVHeatConductionTimeDerivative";
+    bool use_functors = isParamValid("density_functor") || isParamValid("specific_heat_functor");
```
```suggestion
    const bool use_functors = isParamValid("density_functor") || isParamValid("specific_heat_functor");
```

lindsayad commented on modules/heat_transfer/test/tests/physics/tests:
```diff
@@ -54,6 +54,13 @@
+      input = 'test_fv_functor.i'
+      exodiff = 'test_fv_functor_all_out.e'
+      cli_args = "Outputs/file_base=test_fv_functor_all_out"
+      detail = 'with all material properties defined using functors,'
```
```suggestion
      detail = 'with all properties defined using functors,'
```

lindsayad commented on modules/navier_stokes/include/functormaterials/GeneralFunctorFluidProps.h:
```diff
@@ -65,6 +65,15 @@ class GeneralFunctorFluidPropsTempl : public FunctorMaterial,
   const bool _neglect_derivatives_of_density_time_derivative;
 
+  /// Name to use for density functor
+  const MooseFunctorName _density_name;
+  /// Name to use for the dynamic viscosity functor
+  const MooseFunctorName _dynamic_viscosity_name;
+  /// Name to use for the specific heat functor
+  const MooseFunctorName _specific_heat_name;
+  /// Name to use for the thermal conductivity functor
+  const MooseFunctorName _thermal_conductivity_name;
```
```suggestion
  const MooseFunctorName & _density_name;
  /// Name to use for the dynamic viscosity functor
  const MooseFunctorName & _dynamic_viscosity_name;
  /// Name to use for the specific heat functor
  const MooseFunctorName & _specific_heat_name;
  /// Name to use for the thermal conductivity functor
  const MooseFunctorName & _thermal_conductivity_name;
```

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2880538/), step Docs: sync website on 9b41bd2 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/30550/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2880585/), step Generate coverage on 9b41bd2 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>623694</th>
      <th colspan=4>#30550 9b41bd</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.48%</td>
      <td>85.48%</td>
      <td>-0.00%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>112669</td>
      <td>112668</td>
      <td>-1</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>19144</td>
      <td>19145</td>
      <td>+1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/30550/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/30550/coverage/framework)


## Modules coverage

### Heat transfer

<table>
  <thead>
    <tr>
      <th></th>
      <th>623694</th>
      <th colspan=4>#30550 9b41bd</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>89.48%</td>
      <td>89.52%</td>
      <td>+0.04%</td>
      <td>97.22%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>4499</td>
      <td>4528</td>
      <td>+29</td>
      <td>35</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>529</td>
      <td>530</td>
      <td>+1</td>
      <td>1</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/30550/coverage_diff/heat_transfer)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/30550/coverage/heat_transfer)

### Navier stokes

<table>
  <thead>
    <tr>
      <th></th>
      <th>623694</th>
      <th colspan=4>#30550 9b41bd</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>86.46%</td>
      <td>86.46%</td>
      <td>+0.01%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>18420</td>
      <td>18429</td>
      <td>+9</td>
      <td>10</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>2885</td>
      <td>2885</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/30550/coverage_diff/navier_stokes)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/30550/coverage/navier_stokes)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/30550/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/30550/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/30550/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/30550/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/30550/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/30550/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/30550/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/30550/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/30550/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/30550/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/30550/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/30550/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/30550/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/30550/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/30550/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/30550/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/30550/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/30550/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/30550/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/30550/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/30550/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/30550/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/30550/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/30550/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/30550/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/30550/coverage/stochastic_tools)
- [`subchannel`](https://mooseframework.inl.gov/docs/PRs/30550/coverage/subchannel)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/30550/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/30550/coverage/xfem)

</details>



This comment will be updated on new commits.

GiudGiud committed:
Address Alex's review

- add docstring to new kernel
- better distinction between functor and traditional matprops
- use FVFunctorTimekernel
- more references

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2881835/), step Clang format on 062fc38 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/30550/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/30550/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 083e818fdf10367a54e8ba2444e70a9b273db34b`

GiudGiud commented:
Thanks for the review!
